org/mozilla/jss/nss/PR.java

```diff
@@ -66,20 +66,38 @@ public static native PRFDProxy NewBufferPRFD(BufferProxy read_buf,
                                                  BufferProxy write_buf,
                                                  byte[] peer_info);
     /**
-     * Close an existing PRFDProxy.
+     * Close an existing PRFDProxy, clearing the pointer if successful.
      *
      * See also: PR_Close in /usr/include/nspr4/prio.h
      */
-    public static native int Close(PRFDProxy fd);
+    public static int Close(PRFDProxy fd) {
+        return Close(fd, true);
+    }
+
+    /**
+     * Close an existing PRFDProxy with an option to clear the pointer.
+     *
+     * See also: PR_Close in /usr/include/nspr4/prio.h
+     */
+    public static native int Close(PRFDProxy fd, boolean clear);
 
     /**
      * Close an existing SSLFDProxy.
      *
      * See also: org.mozilla.jss.nss.PR.Close
-     *           org.mozilla.jss.nss.SSLFD.releaseNativeResources
+     *           org.mozilla.jss.nss.SSLFDProxy.releaseNativeResources
      */
-    public static int Close(SSLFDProxy fd) throws Exception {
-        int ret = PR.Close((PRFDProxy) fd);
+    public synchronized static int Close(SSLFDProxy fd) throws Exception {
```

Comment:
How about we convert this method into SSLFDProxy.close()?

```
class PRFDProxy extends NativeProxy {
    public void close() {
        PR.Close(this); // don't call JSS_clearPtrFromProxy()
        super.close(); // release native resources
    }
}

class SSLFDProxy extends PRFDProxy {
    public void close() {
        if (!fd.isNull()) {
            SSL.RemoveCallbacks(fd); 
        }
        super.close();
    }
}
```

Reply:
`PR` is static class and contains wrappers for NSPR functions beginning with the prefix `PR_`, of which `PR_Close()` is one. Since `SSLFDProxy` has the same native pointer type as `PRFDProxy` (a `PRFileDesc *` -- the difference is the former has been wrapped by a call to `SSL.ImportFD(...)`) there is no difference. All `PR_` functions work, because NSS creates another `PRFileDesc` layer for SSL, on top of the existing one. We need to do a little more work in `PR_Close` on a `SSLFDProxy` instance, because we need to invoke the `SSLFDProxy.close(...)` method as well, which we don't necessarily want to do if it is `NULL`. 

On the other hand `PRProxy.close()` and `SSLFDProxy.close()` take no arguments and are part of the `NativeProxy`+`AutoCloseable` interface, which free the proxy and clear the `mPointer` instance.

----

Our use of `PR_Close(PRFileDesc *)` is limited mostly to `SSLFDProxy` instances, but we do test it on files as well, in the test suite.

Reply:
Namely, `PRFDProxy` doesn't invoke anything actually -- but we explicitly call `PR.Close(...)` in SSLEngine's cleanup() handler. The issue is we can't call this strictly a "free" method -- closing a `PRFDProxy` via `PR_Close(...)` might send data over the network if the TLS connection hasn't properly terminated. We don't necessarily want to do that (nor do we necessary have space in our buffers, ... &c), hence why `PRFDProxy.close()` doesn't do anything and why `SSLFDProxy.close()` only (tries to) free the `GlobalRefProxy` self-pointer.

Reply:
Keep in mind, `nss.PR` is meant to be a thin wrapper over NSPR functions. Just like `nss.SSL` is a thin wrapper over `SSL_` functions. We're likely going to move as much of this behind a single, large JNI call in an optimized SSLEngine implementation. Some of these details will thus go away.

Reply:
I'm not really familiar with all the details, so I'll let you decide :) And feel free to merge as is since it's already approved. Anything we discussed here is for a possible future enhancement.

To my understanding NSPR only has a `PR_Close(PRFileDesc*)` funcation which is already wrapped by `PR.Close(PRFDProxy)`. The `PR.Close(SSLFDProxy)` on the other hand does not have a corresponding native code in NSPR. Also, that method has a dependency on `SSL` class which is a wrapper for NSS. From design perspective I don't think a wrapper for NSPR should depend on a wrapper for NSS.

I think if we want to do something similar to `PR.Close(PRFDProxy)`, we probably should add an `SSL.Close(SSLFDProxy)` instead. Or we can just add an `SSLFDProxy.close()` like I described earlier. If we don't want to add `PRFDProxy.close()`, we probably can do something like this:

```
class SSLFDProxy extends PRFDProxy {
    public void close() {
        if (fd.isNull()) {
            return;
        }
        SSL.RemoveCallbacks(fd); 
        PR.Close(this, false);
        releaseNativeResources();
    }
}
```
It should be functionally identical to your code, but it will not create an NSPR -> NSS dependency issue.